### ISSUE TITLE
Accuracy issue with MALSD decoding with LM fusion

### ISSUE THREAD

**Issue by YuriiMytiai** [NVIDIA-NeMo/NeMo#14907]
**Describe the bug**

When using MALSD decoding with external N-gram LM fusion for Parakeet models, WER sometimes decreases overall but shows sporadic high deletion rates for certain utterances in the SLURP test set.
In particular, decoding with LM fusion (using `rnnt_decoding.strategy="malsd_batch"`) occasionally removes entire phrases or outputs empty or semantically unrelated hypotheses, while the same model without LM fusion performs normally.
This issue appears across different Parakeet variants and LM weights, but is most pronounced for **parakeet-tdt-0.6b-v2** model at ngram_lm_alpha values ≥ 0.1.

**Steps/Code to reproduce bug**

I've followed [original PR](https://github.com/NVIDIA-NeMo/NeMo/pull/12729) steps to prepare ngram LM and run evaluation.
1. Clone the latest NeMo main branch (commit 1469922d88eebe5e2d498e4a56277cf53505fd5a)
2. Install all the dependencies
3. Train LM:
```
python3 scripts/asr_language_modeling/ngram_lm/train_kenlm.py \
    nemo_model_file="work_dir/parakeet-tdt-0.6b-v2.nemo" \
    train_paths=["work_dir/slurp_train.txt"] \
    kenlm_bin_path="/home/ubuntu/kenlm/build/bin" \
    kenlm_model_file="work_dir/slurp/lm-o6-parakeet_tdt_v2.arpa" \
    ngram_length=6 preserve_arpa=true save_nemo=true
```
4. Run decoding with MALSD + LM fusion:
```
python3 examples/asr/speech_to_text_eval.py \
    model_path="work_dir/parakeet-tdt-0.6b-v2.nemo" \
    dataset_manifest="work_dir/slurp_test_manifest.jsonl" \
    output_filename="work_dir/slurp_test-parakeet_tdt_v2-lmw01.jsonl" \
    batch_size=64 \
    rnnt_decoding.strategy="malsd_batch" \
    rnnt_decoding.beam.max_symbols_per_step=10 \
    rnnt_decoding.beam.beam_size=12 \
    rnnt_decoding.beam.ngram_lm_model="work_dir/slurp/lm-o6-parakeet_tdt_v2.arpa.nemo" \
    rnnt_decoding.beam.ngram_lm_alpha=0.1 \
    text_processing.rm_punctuation=true text_processing.do_lowercase=true
```

**Observed behavior**

While the overall WER often improves with LM fusion as presented in [paper](https://arxiv.org/pdf/2011.13205) and [original PR](https://github.com/NVIDIA-NeMo/NeMo/pull/12729), certain utterances experience severe deletions or semantic drift.
Here's accuracy vs. LM weight results for different Parakeet models. Here **Parakeet-v2 orig-texts** stands for fusion with LM trained on original SRULM train texts (lowercase without punctuation) and **Parakeet-v2 PnC-texts** stands for fusion with LM trained on pre-processed SRULM train texts where preprocessing includes applying PunctuationCapitalizationModel in order to LM match expected ASR model outputs for **parakeet-tdt-0.6b-v2** model:
<img width="944" height="155" alt="Image" src="https://github.com/user-attachments/assets/0d84b06f-ddbf-46d1-b004-8fb9a802fd40" />

As you can see from the table, the numbers highlighted with green correspond to LM weight 0.1 and shows the lowest Deletions rate, as well as lower WER overall, for **parakeet-tdt-0.6b-v2**. But at the same time, if we explore predictions for particular test segments we can find that in some cases hypothesis with LM fusion contains spontaneous deletions, usually at that the beginning/ending of segment:
<img width="1039" height="163" alt="Image" src="https://github.com/user-attachments/assets/604f8571-6aa0-4692-be6b-ec19b3b19071" />

This effect persists with other models as well, e.g. for **parakeet-tdt-1.1b** with LM weight 0.2:
<img width="1038" height="122" alt="Image" src="https://github.com/user-attachments/assets/9658f1b0-71a2-442d-b951-ee79e332a448" />

This effect is even more noticeable if audio segments are relatively long (20-40 sec), which is not a case for SRULP dataset, but is common case for ASR batched decoding.

**Expected behavior**

LM fusion should not degrade hypothesis completeness.
Expected: minor corrections or improved lexical accuracy, not catastrophic deletions.
Observed: utterances reduced to fragments when LM fusion is active.

**Environment overview**

Environment location: Oracle Cloud (shape VM.GPU.A10.1)
Method of NeMo install: from source (pip install -e .)
NeMo commit: 1469922d88eebe5e2d498e4a56277cf53505fd5a (Oct 6 2025, latest main)

**Environment details**

OS: Ubuntu 22.04.5 LTS
GPU: NVIDIA A10
Driver: 550.163.01
CUDA: 12.4
Python: 3.10.12
PyTorch: 2.7.1 + cu124

**Additional context**

The issue reproduces across Parakeet-rnnt-1.1b, Parakeet-tdt-1.1b, and Parakeet-TDT-0.6b-v2 models, but is most pronounced on Parakeet-TDT-0.6b-v2. Also, the issue reproduces across different datasets, whether they are used for ASR and/or LM model training, or accuracy evaluation.
Training corpus for LM: SLURP train set, optionally preprocessed with punctuation-capitalization model for Parakeet-TDT-0.6b-v2 to match its output style.
Hardware appears unrelated — reproducible across different GPUs.

**Comment by lilithgrigoryan** [comment]
During language model fusion, the output probabilities from an ASR model are combined with probabilities from an n-gram language model using a weighted sum. At each decoding step, tokens are selected based on these combined probabilities. However, this combination of ASR and LM scores does not always guarantee improved transcripts, especially when the LM is trained on a relatively small or non-diverse text corpus. In some cases, fusion can improve transcription accuracy, while in others it may degrade it. Overall improvement on a test set indicates that decoding with the LM is effective in your scenario.

Higher deletion rates can occur because the LM, by design, does not account for blank tokens in transducer models and does not assign additional probability to them. This can lead to more frequent blank predictions, resulting in increased deletions. To mitigate this, we apply a rescoring technique known as blank weighting, which reduces the impact of this issue, but of course sometimes you may see deletions.

**Comment by nithinraok** [comment]
As there has been no response, I am closing this as completed. Please feel free to reopen it if the suggested info is incomplete.
